pkg/pillar/zedcloud/deferred.go

```diff
@@ -66,55 +84,81 @@ func HandleDeferred(zedcloudCtx *ZedCloudContext, event time.Time, spacing time.
 		event, spacing, len(ctx.deferredItems))
-	iteration := 0 // Do some load spreading
-	for key, l := range ctx.deferredItems {
-		log.Functionf("Trying to send for %s items %d\n", key, len(l.list))
-		failed := false
-		for i, item := range l.list {
-			if item.buf == nil {
-				continue
-			}
-			if item.buf.Len() == 0 {
-				log.Errorf("Zero length defered item for %s",
-					key)
-				continue
+
+	//sort items by added time in ascending order
+	sort.Slice(ctx.deferredItems, func(i, j int) bool {
+		t1 := time.Time{}
+		for _, el := range ctx.deferredItems[i].list {
+			if t1.IsZero() || el.added.Before(t1) {
+				t1 = el.added
 			}
-			log.Functionf("Trying to send for %s item %d data size %d\n",
-				key, i, item.size)
-			resp, _, _, err := SendOnAllIntf(item.zedcloudCtx, item.url,
-				item.size, item.buf, iteration, item.bailOnHTTPErr)
-			if item.bailOnHTTPErr && resp != nil &&
-				resp.StatusCode >= 400 && resp.StatusCode < 600 {
-				log.Functionf("HandleDeferred: for %s ignore code %d\n",
-					key, resp.StatusCode)
-			} else if err != nil {
-				log.Functionf("HandleDeferred: for %s failed %s\n",
-					key, err)
-				failed = true
-				break
+		}
+		t2 := time.Time{}
+		for _, el := range ctx.deferredItems[j].list {
+			if t2.IsZero() || el.added.Before(t2) {
+				t2 = el.added
 			}
-			item.buf = nil
 		}
-		if failed {
-			break
-		} else {
-			delete(ctx.deferredItems, key)
-			iteration += 1
-			// XXX sleeping in main thread
-			if len(ctx.deferredItems) != 0 && spacing != 0 {
-				log.Functionf("HandleDeferred sleeping %v\n",
-					spacing)
-				time.Sleep(spacing)
+		return t1.Before(t2)
+	})
+
+	iteration := 0 // Do some load spreading
+topLoop:
+	for _, f := range ctx.priorityCheckFunctions {
+		for ind := 0; ind < len(ctx.deferredItems); ind++ {
+			key := ctx.deferredItems[ind].key
+			l := ctx.deferredItems[ind]
+			log.Functionf("handleDeferred: Trying to send for %s items %d\n", key, len(l.list))
+			if f(l.itemType) {
+				for i, item := range l.list {
+					if item.buf == nil {
+						continue
+					}
+					if item.buf.Len() == 0 {
+						log.Functionf("handleDeferred: Zero length deferred item for %s",
+							key)
+						continue
+					}
+					log.Functionf("handleDeferred: Trying to send for %s item %d data size %d\n",
+						key, i, item.size)
+					resp, _, _, err := SendOnAllIntf(item.zedcloudCtx, item.url,
+						item.size, item.buf, iteration, item.bailOnHTTPErr)
+					if item.bailOnHTTPErr && resp != nil &&
+						resp.StatusCode >= 400 && resp.StatusCode < 600 {
+						log.Functionf("handleDeferred: for %s ignore code %d\n",
+							key, resp.StatusCode)
+					} else if err != nil {
+						log.Functionf("handleDeferred: for %s failed %s\n",
+							key, err)
+						break topLoop
+					}
+					item.buf = nil
+				}
+				ctx.deferredItems = append(ctx.deferredItems[:ind], ctx.deferredItems[ind+1:]...)
+				ind--
+				iteration++
+				// XXX sleeping in main thread
```

Comment:
Given that you call this from different places, how do we know that this doesn't cause an unbounded delay/execution time when the network is slow or partially failing.

Reply:
I replace direct call to Handle with triggering the timer event.
Also I add some kind of time checking inside handleDeferred to stop loop on long delays.

Reply:
OK